File: apps/desktop/src/organizer.ts
import { invoke } from "@tauri-apps/api/core";
import { Task } from "./tasks";

export type ListView = "active" | "archived" | "trash" | "all";
export type GoalKind =
  "outcome" | "skill" | "habit" | "project" | "academic" | "personal";
export type GoalState = "active" | "paused" | "achieved" | "retired";
export type ProjectState =
  | "idea"
  | "exploring"
  | "planned"
  | "active"
  | "paused"
  | "completed"
  | "archived"
  | "abandoned";
export type MilestoneState = "planned" | "in_progress" | "achieved" | "skipped";
export type Blocker = { entity: string; count: number };
export type ProductError = {
  code:
    | "not_found"
    | "revision_conflict"
    | "validation"
    | "assignment_unavailable"
    | "trash_blocked"
    | "unavailable";
  blockers: Blocker[];
};

export function asProductError(reason: unknown): ProductError {
  if (typeof reason === "object" && reason !== null && "code" in reason) {
    const value = reason as Partial<ProductError>;
    const valid = [
      "not_found",
      "revision_conflict",
      "validation",
      "assignment_unavailable",
      "trash_blocked",
      "unavailable",
    ];
    if (typeof value.code === "string" && valid.includes(value.code)) {
      return {
        code: value.code as ProductError["code"],
        blockers: Array.isArray(value.blockers) ? value.blockers : [],
      };
    }
  }
  return { code: "unavailable", blockers: [] };
}

type Canonical = {
  id: string;
  created_at: string;
  updated_at: string;
  revision: number;
  trashed_at: string | null;
};
export type Area = Canonical & {
  name: string;
  description: string | null;
  archived_at: string | null;
};
export type Goal = Canonical & {
  area_id: string | null;
  title: string;
  description: string | null;
  kind: GoalKind;
  state: GoalState;
  archived_at: string | null;
};
export type Project = Canonical & {
  goal_id: string | null;
  title: string;
  description: string | null;
  state: ProjectState;
  completed_at: string | null;
  archived_at: string | null;
};
type Milestone = Canonical & {
  title: string;
  state: MilestoneState;
  target_date: string | null;
  achieved_at: string | null;
  position: number;
};
export type GoalMilestone = Milestone & { goal_id: string };
export type ProjectMilestone = Milestone & { project_id: string };
export type AreaDetail = { area: Area; goals: Goal[] };
export type GoalSummary = {
  milestone_total: number;
  milestone_achieved: number;
  project_total: number;
  task_total: number;
  task_completed: number;
};
export type GoalDetail = {
  goal: Goal;
  summary: GoalSummary;
  milestones: GoalMilestone[];
  projects: Project[];
  direct_tasks: Task[];
  project_tasks: Task[];
};
export type ProjectSummary = {
  milestone_total: number;
  milestone_achieved: number;
  task_total: number;
  task_completed: number;
};
export type Activity = {
  event_id: string;
  occurred_at: string;
  entity_type: "project" | "project_milestone";
  entity_id: string;
  action: string;
  from_revision: number | null;
  to_revision: number | null;
  command_id: string;
};
export type ProjectDetail = {
  project: Project;
  summary: ProjectSummary;
  milestones: ProjectMilestone[];
  current_milestone: ProjectMilestone | null;
  tasks: Task[];
  next_actions: Task[];
  recent_activity: Activity[];
};

const revision = (item: Pick<Canonical, "revision">) => ({
  expected_revision: item.revision,
});

export const areaClient = {
  list: (view: ListView = "all") => invoke<Area[]>("list_areas", { view }),
  get: (area: Pick<Area, "id">) =>
    invoke<AreaDetail>("get_area", { areaId: area.id }),
  create: (input: { name: string; description: string | null }) =>
    invoke<Area>("create_area", { input }),
  update: (area: Area, input: { name?: string; description?: string | null }) =>
    invoke<Area>("update_area", {
      areaId: area.id,
      input: { ...input, expected_revision: area.revision },
    }),
  archive: (area: Area) =>
    invoke<Area>("archive_area", { areaId: area.id, input: revision(area) }),
  unarchive: (area: Area) =>
    invoke<Area>("unarchive_area", { areaId: area.id, input: revision(area) }),
  trash: (area: Area) =>
    invoke<Area>("trash_area", { areaId: area.id, input: revision(area) }),
  restore: (area: Pick<Area, "id" | "revision">) =>
    invoke<Area>("restore_area", { areaId: area.id, input: revision(area) }),
};

export const goalClient = {
  list: (view: ListView = "all") => invoke<Goal[]>("list_goals", { view }),
  get: (goal: Pick<Goal, "id">) =>
    invoke<GoalDetail>("get_goal_detail", { goalId: goal.id }),
  create: (input: {
    title: string;
    description: string | null;
    kind: GoalKind;
    area_id: string | null;
  }) => invoke<Goal>("create_goal", { input }),
  update: (
    goal: Goal,
    input: { title?: string; description?: string | null; kind?: GoalKind },
  ) =>
    invoke<Goal>("update_goal", {
      goalId: goal.id,
      input: { ...input, expected_revision: goal.revision },
    }),
  setState: (goal: Goal, state: GoalState) =>
    invoke<Goal>("set_goal_state", {
      goalId: goal.id,
      input: { state, expected_revision: goal.revision },
    }),
  setArea: (goal: Goal, area_id: string | null) =>
    invoke<Goal>("set_goal_area", {
      goalId: goal.id,
      input: { area_id, expected_revision: goal.revision },
    }),
  archive: (goal: Goal) =>
    invoke<Goal>("archive_goal", { goalId: goal.id, input: revision(goal) }),
  unarchive: (goal: Goal) =>
    invoke<Goal>("unarchive_goal", { goalId: goal.id, input: revision(goal) }),
  trash: (goal: Goal) =>
    invoke<Goal>("trash_goal", { goalId: goal.id, input: revision(goal) }),
  restore: (goal: Pick<Goal, "id" | "revision">) =>
    invoke<Goal>("restore_goal", { goalId: goal.id, input: revision(goal) }),
};

export const projectClient = {
  list: (view: ListView = "all") =>
    invoke<Project[]>("list_projects", { view }),
  get: (project: Pick<Project, "id">) =>
    invoke<ProjectDetail>("get_project_detail", { projectId: project.id }),
  create: (input: {
    title: string;
    description: string | null;
    state: ProjectState;
    goal_id: string | null;
  }) => invoke<Project>("create_project", { input }),
  update: (
    project: Project,
    input: { title?: string; description?: string | null },
  ) =>
    invoke<Project>("update_project", {
      projectId: project.id,
      input: { ...input, expected_revision: project.revision },
    }),
  setState: (project: Project, state: Exclude<ProjectState, "archived">) =>
    invoke<Project>("set_project_state", {
      projectId: project.id,
      input: { state, expected_revision: project.revision },
    }),
  setGoal: (project: Project, goal_id: string | null) =>
    invoke<Project>("set_project_goal", {
      projectId: project.id,
      input: { goal_id, expected_revision: project.revision },
    }),
  archive: (project: Project) =>
    invoke<Project>("archive_project", {
      projectId: project.id,
      input: revision(project),
    }),
  unarchive: (project: Project) =>
    invoke<Project>("unarchive_project", {
      projectId: project.id,
      input: revision(project),
    }),
  trash: (project: Project) =>
    invoke<Project>("trash_project", {
      projectId: project.id,
      input: revision(project),
    }),
  restore: (project: Pick<Project, "id" | "revision">) =>
    invoke<Project>("restore_project", {
      projectId: project.id,
      input: revision(project),
    }),
};

type MilestoneInput = { title: string; target_date: string | null };
const reorderInput = (items: Milestone[]) => ({
  items: items.map((item) => ({
    id: item.id,
    expected_revision: item.revision,
  })),
});

export const goalMilestoneClient = {
  list: (goal: Goal, trashed = false) =>
    invoke<GoalMilestone[]>("list_goal_milestones", {
      goalId: goal.id,
      trashed,
    }),
  create: (goal: Goal, input: MilestoneInput) =>
    invoke<GoalMilestone>("create_goal_milestone", { goalId: goal.id, input }),
  update: (milestone: GoalMilestone, input: Partial<MilestoneInput>) =>
    invoke<GoalMilestone>("update_goal_milestone", {
      milestoneId: milestone.id,
      input: { ...input, expected_revision: milestone.revision },
    }),
  setState: (milestone: GoalMilestone, state: MilestoneState) =>
    invoke<GoalMilestone>("set_goal_milestone_state", {
      milestoneId: milestone.id,
      input: { state, expected_revision: milestone.revision },
    }),
  reorder: (goal: Goal, items: GoalMilestone[]) =>
    invoke<GoalMilestone[]>("reorder_goal_milestones", {
      goalId: goal.id,
      input: reorderInput(items),
    }),
  trash: (milestone: GoalMilestone) =>
    invoke<GoalMilestone>("trash_goal_milestone", {
      milestoneId: milestone.id,
      input: revision(milestone),
    }),
  restore: (milestone: Pick<GoalMilestone, "id" | "revision">) =>
    invoke<GoalMilestone>("restore_goal_milestone", {
      milestoneId: milestone.id,
      input: revision(milestone),
    }),
};

export const projectMilestoneClient = {
  list: (project: Project, trashed = false) =>
    invoke<ProjectMilestone[]>("list_project_milestones", {
      projectId: project.id,
      trashed,
    }),
  create: (project: Project, input: MilestoneInput) =>
    invoke<ProjectMilestone>("create_project_milestone", {
      projectId: project.id,
      input,
    }),
  update: (milestone: ProjectMilestone, input: Partial<MilestoneInput>) =>
    invoke<ProjectMilestone>("update_project_milestone", {
      milestoneId: milestone.id,
      input: { ...input, expected_revision: milestone.revision },
    }),
  setState: (milestone: ProjectMilestone, state: MilestoneState) =>
    invoke<ProjectMilestone>("set_project_milestone_state", {
      milestoneId: milestone.id,
      input: { state, expected_revision: milestone.revision },
    }),
  reorder: (project: Project, items: ProjectMilestone[]) =>
    invoke<ProjectMilestone[]>("reorder_project_milestones", {
      projectId: project.id,
      input: reorderInput(items),
    }),
  trash: (milestone: ProjectMilestone) =>
    invoke<ProjectMilestone>("trash_project_milestone", {
      milestoneId: milestone.id,
      input: revision(milestone),
    }),
  restore: (milestone: Pick<ProjectMilestone, "id" | "revision">) =>
    invoke<ProjectMilestone>("restore_project_milestone", {
      milestoneId: milestone.id,
      input: revision(milestone),
    }),
};

File: apps/desktop/src-tauri/src/organizer.rs
//! Fixed Phase 1B organizer commands. The authenticated request primitive stays Rust-only.

use serde::{Deserialize, Serialize};
use tauri::State;

use crate::service::{
    product_request, ProductError, ProductErrorCode, RevisionInput, ServiceState, Task,
};

#[derive(Debug, Clone, Deserialize, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ListView {
    Active,
    Archived,
    Trash,
    All,
}

impl ListView {
    fn value(&self) -> &'static str {
        match self {
            Self::Active => "active",
            Self::Archived => "archived",
            Self::Trash => "trash",
            Self::All => "all",
        }
    }
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AreaCreateInput {
    pub name: String,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AreaUpdateInput {
    pub expected_revision: i64,
    pub name: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Area {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub archived_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    pub trashed_at: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AreaDetail {
    pub area: Area,
    pub goals: Vec<Goal>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalCreateInput {
    pub title: String,
    pub description: Option<String>,
    pub kind: String,
    pub area_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalUpdateInput {
    pub expected_revision: i64,
    pub title: Option<String>,
    pub description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub kind: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalStateInput {
    pub expected_revision: i64,
    pub state: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalAreaInput {
    pub expected_revision: i64,
    pub area_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Goal {
    pub id: String,
    pub area_id: Option<String>,
    pub title: String,
    pub description: Option<String>,
    pub kind: String,
    pub state: String,
    pub archived_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    pub trashed_at: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalSummary {
    pub milestone_total: i64,
    pub milestone_achieved: i64,
    pub project_total: i64,
    pub task_total: i64,
    pub task_completed: i64,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalDetail {
    pub goal: Goal,
    pub summary: GoalSummary,
    pub milestones: Vec<GoalMilestone>,
    pub projects: Vec<Project>,
    pub direct_tasks: Vec<Task>,
    pub project_tasks: Vec<Task>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectCreateInput {
    pub title: String,
    pub description: Option<String>,
    pub state: String,
    pub goal_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectUpdateInput {
    pub expected_revision: i64,
    pub title: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectStateInput {
    pub expected_revision: i64,
    pub state: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectGoalInput {
    pub expected_revision: i64,
    pub goal_id: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Project {
    pub id: String,
    pub goal_id: Option<String>,
    pub title: String,
    pub description: Option<String>,
    pub state: String,
    pub completed_at: Option<String>,
    pub archived_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    pub trashed_at: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectSummary {
    pub milestone_total: i64,
    pub milestone_achieved: i64,
    pub task_total: i64,
    pub task_completed: i64,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Activity {
    pub event_id: String,
    pub occurred_at: String,
    pub entity_type: String,
    pub entity_id: String,
    pub action: String,
    pub from_revision: Option<i64>,
    pub to_revision: Option<i64>,
    pub command_id: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RecoveryItem {
    pub entity_type: String,
    pub entity_id: String,
    pub label: String,
    pub lifecycle: String,
    pub revision: i64,
    pub trashed_at: String,
    pub owner_label: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RecoveryActivity {
    pub event_id: String,
    pub occurred_at: String,
    pub entity_type: String,
    pub entity_id: String,
    pub label: String,
    pub action: String,
    pub authority: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct RecoveryOutput {
    pub trash: Vec<RecoveryItem>,
    pub recent_activity: Vec<RecoveryActivity>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectDetail {
    pub project: Project,
    pub summary: ProjectSummary,
    pub milestones: Vec<ProjectMilestone>,
    pub current_milestone: Option<ProjectMilestone>,
    pub tasks: Vec<Task>,
    pub next_actions: Vec<Task>,
    pub recent_activity: Vec<Activity>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MilestoneCreateInput {
    pub title: String,
    pub target_date: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MilestoneUpdateInput {
    pub expected_revision: i64,
    pub title: Option<String>,
    pub target_date: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct MilestoneStateInput {
    pub expected_revision: i64,
    pub state: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ReorderMilestoneItem {
    pub id: String,
    pub expected_revision: i64,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ReorderMilestonesInput {
    pub items: Vec<ReorderMilestoneItem>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GoalMilestone {
    pub id: String,
    pub goal_id: String,
    pub title: String,
    pub state: String,
    pub target_date: Option<String>,
    pub achieved_at: Option<String>,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    pub trashed_at: Option<String>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct ProjectMilestone {
    pub id: String,
    pub project_id: String,
    pub title: String,
    pub state: String,
    pub target_date: Option<String>,
    pub achieved_at: Option<String>,
    pub position: i64,
    pub created_at: String,
    pub updated_at: String,
    pub revision: i64,
    pub trashed_at: Option<String>,
}

fn entity_route(prefix: &str, id: &str, suffix: &str) -> Result<String, ProductError> {
    let valid = id.len() == 36
        && id.bytes().enumerate().all(|(index, byte)| match index {
            8 | 13 | 18 | 23 => byte == b'-',
            _ => byte.is_ascii_hexdigit(),
        });
    if !valid {
        return Err(ProductError::new(ProductErrorCode::Validation));
    }
    Ok(format!("{prefix}/{id}{suffix}"))
}

macro_rules! list_command {
    ($name:ident, $route:literal, $output:ty) => {
        #[tauri::command]
        pub async fn $name(
            state: State<'_, ServiceState>,
            view: ListView,
        ) -> Result<Vec<$output>, ProductError> {
            let route = format!(concat!($route, "?view={}"), view.value());
            product_request::<(), Vec<$output>>(&state, reqwest::Method::GET, &route, None).await
        }
    };
}

list_command!(list_areas, "/v1/areas", Area);
list_command!(list_goals, "/v1/goals", Goal);
list_command!(list_projects, "/v1/projects", Project);

#[tauri::command]
pub async fn get_recovery(state: State<'_, ServiceState>) -> Result<RecoveryOutput, ProductError> {
    product_request::<(), RecoveryOutput>(&state, reqwest::Method::GET, "/v1/recovery", None).await
}

macro_rules! create_command {
    ($name:ident, $route:literal, $input:ty, $output:ty) => {
        #[tauri::command]
        pub async fn $name(
            state: State<'_, ServiceState>,
            input: $input,
        ) -> Result<$output, ProductError> {
            product_request(&state, reqwest::Method::POST, $route, Some(&input)).await
        }
    };
}

create_command!(create_area, "/v1/areas", AreaCreateInput, Area);
create_command!(create_goal, "/v1/goals", GoalCreateInput, Goal);
create_command!(create_project, "/v1/projects", ProjectCreateInput, Project);

#[tauri::command]
pub async fn get_area(
    state: State<'_, ServiceState>,
    area_id: String,
) -> Result<AreaDetail, ProductError> {
    let route = entity_route("/v1/areas", &area_id, "")?;
    product_request::<(), AreaDetail>(&state, reqwest::Method::GET, &route, None).await
}

#[tauri::command]
pub async fn get_goal_detail(
    state: State<'_, ServiceState>,
    goal_id: String,
) -> Result<GoalDetail, ProductError> {
    let route = entity_route("/v1/goals", &goal_id, "")?;
    product_request::<(), GoalDetail>(&state, reqwest::Method::GET, &route, None).await
}

#[tauri::command]
pub async fn get_project_detail(
    state: State<'_, ServiceState>,
    project_id: String,
) -> Result<ProjectDetail, ProductError> {
    let route = entity_route("/v1/projects", &project_id, "")?;
    product_request::<(), ProjectDetail>(&state, reqwest::Method::GET, &route, None).await
}

macro_rules! update_command {
    ($name:ident, $prefix:literal, $id_name:ident, $input:ty, $output:ty) => {
        #[tauri::command]
        pub async fn $name(
            state: State<'_, ServiceState>,
            $id_name: String,
            input: $input,
        ) -> Result<$output, ProductError> {
            let route = entity_route($prefix, &$id_name, "")?;
            product_request(&state, reqwest::Method::PATCH, &route, Some(&input)).await
        }
    };
}

update_command!(update_area, "/v1/areas", area_id, AreaUpdateInput, Area);
update_command!(update_goal, "/v1/goals", goal_id, GoalUpdateInput, Goal);
update_command!(
    update_project,
    "/v1/projects",
    project_id,
    ProjectUpdateInput,
    Project
);
update_command!(
    update_goal_milestone,
    "/v1/goal-milestones",
    milestone_id,
    MilestoneUpdateInput,
    GoalMilestone
);
update_command!(
    update_project_milestone,
    "/v1/project-milestones",
    milestone_id,
    MilestoneUpdateInput,
    ProjectMilestone
);

async fn revision_action<R: for<'de> Deserialize<'de>>(
    state: &ServiceState,
    prefix: &str,
    entity_id: &str,
    action: &str,
    input: &RevisionInput,
) -> Result<R, ProductError> {
    let route = entity_route(prefix, entity_id, &format!("/{action}"))?;
    product_request(state, reqwest::Method::POST, &route, Some(input)).await
}

macro_rules! revision_command {
    ($name:ident, $prefix:literal, $id_name:ident, $action:literal, $output:ty) => {
        #[tauri::command]
        pub async fn $name(
            state: State<'_, ServiceState>,
            $id_name: String,
            input: RevisionInput,
        ) -> Result<$output, ProductError> {
            revision_action(&state, $prefix, &$id_name, $action, &input).await
        }
    };
}

revision_command!(archive_area, "/v1/areas", area_id, "archive", Area);
revision_command!(unarchive_area, "/v1/areas", area_id, "unarchive", Area);
revision_command!(trash_area, "/v1/areas", area_id, "trash", Area);
revision_command!(restore_area, "/v1/areas", area_id, "restore", Area);
revision_command!(archive_goal, "/v1/goals", goal_id, "archive", Goal);
revision_command!(unarchive_goal, "/v1/goals", goal_id, "unarchive", Goal);
revision_command!(trash_goal, "/v1/goals", goal_id, "trash", Goal);
revision_command!(restore_goal, "/v1/goals", goal_id, "restore", Goal);
revision_command!(
    archive_project,
    "/v1/projects",
    project_id,
    "archive",
    Project
);
revision_command!(
    unarchive_project,
    "/v1/projects",
    project_id,
    "unarchive",
    Project
);
revision_command!(trash_project, "/v1/projects", project_id, "trash", Project);
revision_command!(
    restore_project,
    "/v1/projects",
    project_id,
    "restore",
    Project
);
revision_command!(
    trash_goal_milestone,
    "/v1/goal-milestones",
    milestone_id,
    "trash",
    GoalMilestone
);
revision_command!(
    restore_goal_milestone,
    "/v1/goal-milestones",
    milestone_id,
    "restore",
    GoalMilestone
);
revision_command!(
    trash_project_milestone,
    "/v1/project-milestones",
    milestone_id,
    "trash",
    ProjectMilestone
);
revision_command!(
    restore_project_milestone,
    "/v1/project-milestones",
    milestone_id,
    "restore",
    ProjectMilestone
);

async fn put_action<T: Serialize, R: for<'de> Deserialize<'de>>(
    state: &ServiceState,
    prefix: &str,
    entity_id: &str,
    action: &str,
    input: &T,
) -> Result<R, ProductError> {
    let route = entity_route(prefix, entity_id, &format!("/{action}"))?;
    product_request(state, reqwest::Method::PUT, &route, Some(input)).await
}

#[tauri::command]
pub async fn set_goal_state(
    state: State<'_, ServiceState>,
    goal_id: String,
    input: GoalStateInput,
) -> Result<Goal, ProductError> {
    put_action(&state, "/v1/goals", &goal_id, "state", &input).await
}

#[tauri::command]
pub async fn set_goal_area(
    state: State<'_, ServiceState>,
    goal_id: String,
    input: GoalAreaInput,
) -> Result<Goal, ProductError> {
    put_action(&state, "/v1/goals", &goal_id, "area", &input).await
}

#[tauri::command]
pub async fn set_project_state(
    state: State<'_, ServiceState>,
    project_id: String,
    input: ProjectStateInput,
) -> Result<Project, ProductError> {
    put_action(&state, "/v1/projects", &project_id, "state", &input).await
}

#[tauri::command]
pub async fn set_project_goal(
    state: State<'_, ServiceState>,
    project_id: String,
    input: ProjectGoalInput,
) -> Result<Project, ProductError> {
    put_action(&state, "/v1/projects", &project_id, "goal", &input).await
}

#[tauri::command]
pub async fn set_goal_milestone_state(
    state: State<'_, ServiceState>,
    milestone_id: String,
    input: MilestoneStateInput,
) -> Result<GoalMilestone, ProductError> {
    put_action(
        &state,
        "/v1/goal-milestones",
        &milestone_id,
        "state",
        &input,
    )
    .await
}

#[tauri::command]
pub async fn set_project_milestone_state(
    state: State<'_, ServiceState>,
    milestone_id: String,
    input: MilestoneStateInput,
) -> Result<ProjectMilestone, ProductError> {
    put_action(
        &state,
        "/v1/project-milestones",
        &milestone_id,
        "state",
        &input,
    )
    .await
}

#[tauri::command]
pub async fn create_goal_milestone(
    state: State<'_, ServiceState>,
    goal_id: String,
    input: MilestoneCreateInput,
) -> Result<GoalMilestone, ProductError> {
    let route = entity_route("/v1/goals", &goal_id, "/milestones")?;
    product_request(&state, reqwest::Method::POST, &route, Some(&input)).await
}

#[tauri::command]
pub async fn create_project_milestone(
    state: State<'_, ServiceState>,
    project_id: String,
    input: MilestoneCreateInput,
) -> Result<ProjectMilestone, ProductError> {
    let route = entity_route("/v1/projects", &project_id, "/milestones")?;
    product_request(&state, reqwest::Method::POST, &route, Some(&input)).await
}

#[tauri::command]
pub async fn reorder_goal_milestones(
    state: State<'_, ServiceState>,
    goal_id: String,
    input: ReorderMilestonesInput,
) -> Result<Vec<GoalMilestone>, ProductError> {
    let route = entity_route("/v1/goals", &goal_id, "/milestones/reorder")?;
    product_request(&state, reqwest::Method::PUT, &route, Some(&input)).await
}

#[tauri::command]
pub async fn reorder_project_milestones(
    state: State<'_, ServiceState>,
    project_id: String,
    input: ReorderMilestonesInput,
) -> Result<Vec<ProjectMilestone>, ProductError> {
    let route = entity_route("/v1/projects", &project_id, "/milestones/reorder")?;
    product_request(&state, reqwest::Method::PUT, &route, Some(&input)).await
}

async fn list_milestones<R: for<'de> Deserialize<'de>>(
    state: &ServiceState,
    prefix: &str,
    owner_id: &str,
    trashed: bool,
) -> Result<Vec<R>, ProductError> {
    let suffix = format!("/milestones?trashed={trashed}");
    let route = entity_route(prefix, owner_id, &suffix)?;
    product_request::<(), Vec<R>>(state, reqwest::Method::GET, &route, None).await
}

#[tauri::command]
pub async fn list_goal_milestones(
    state: State<'_, ServiceState>,
    goal_id: String,
    trashed: bool,
) -> Result<Vec<GoalMilestone>, ProductError> {
    list_milestones(&state, "/v1/goals", &goal_id, trashed).await
}

#[tauri::command]
pub async fn list_project_milestones(
    state: State<'_, ServiceState>,
    project_id: String,
    trashed: bool,
) -> Result<Vec<ProjectMilestone>, ProductError> {
    list_milestones(&state, "/v1/projects", &project_id, trashed).await
}

#[cfg(test)]
mod tests {
    use super::{entity_route, GoalCreateInput, GoalUpdateInput, ListView};

    #[test]
    fn routes_accept_only_canonical_entity_ids() {
        let id = "11111111-1111-4111-8111-111111111111";
        assert_eq!(
            entity_route("/v1/goals", id, "/archive").unwrap(),
            format!("/v1/goals/{id}/archive")
        );
        assert!(entity_route("/v1/goals", "../health", "").is_err());
    }

    #[test]
    fn list_views_and_inputs_serialize_to_backend_contract() {
        assert_eq!(ListView::Archived.value(), "archived");
        let input = GoalCreateInput {
            title: "Synthetic Goal".into(),
            description: None,
            kind: "outcome".into(),
            area_id: None,
        };
        let value = serde_json::to_value(input).unwrap();
        assert_eq!(value["area_id"], serde_json::Value::Null);
        assert_eq!(value["kind"], "outcome");
    }

    #[test]
    fn goal_patch_omits_an_unset_non_nullable_kind() {
        let input = GoalUpdateInput {
            expected_revision: 2,
            title: Some("Edited Goal".into()),
            description: None,
            kind: None,
        };
        let value = serde_json::to_value(input).unwrap();

        assert_eq!(value["description"], serde_json::Value::Null);
        assert!(value.get("kind").is_none());
    }
}
